### PR TITLE
#3245 Feature: Display session details of applicants in CodeTrek.

### PR DESCRIPTION
**Pull request by ImRishabhRawat**
Targets #3245 

## Description for the Session page
Sometimes we conduct sessions related to engineering practices and concept in CodeTrek which are given by CodeTrek candidates and sometimes ColoredCow employees. For CodeTrek Candidates as this data related to their presented sessions is something  that can help in their evalutaion, so we want a functionality to store the presentration data like date, topic and, presentation link and additional information in portal. 

The data should be stored in the database which enables efficient retrieval and management of candidate of this data.

## Breakdown & Estimates 

**Expected Time and Approach**

- [x] Create a table in database and define its structure. `1-2 Hours`
- [x] Define relationship between codetrek_applicant_sessions table and codetrek_applicant table. `1-2 Hours`
- [x] Create a route. `0-1 Hour`
- [x] Define Controller and define function for each task like adding new data, edit, and delete the data. `2-3 Hours`
- [x] Create view file to implement presentation details page. `0-1 Hour`
- [x] Implement add new presentation - Create  a form. `2-3 Hour`
- [x] Read about implementing backend validation (Form Requests). `0-1 Hour`
- [x] Add backend validation. `0-1Hour`
- [x] Show appropriate error in frontend incase any validation rules gets failed. `1-2 Hours`
- [x] Implement edit feature - by adding form. `1-2 Hours`
- [x] Implement sorting. `0-1 Hours`
- [x] Implement delete functionality. `1-2 Hours`

### Total Time : 20-21 Hours

**Actual Time and Approach** 

**Designing**
- [x] Create design of Session page in Figma.  `2-3 Hour`
- [x] Create database schema : Many to Many (Polymorphic) relationship and create a presentation on google slides. `2-3 Hour`

**Coding**
- [x] Create 'sessions' table.  
- [x] Create 'model_has_sessions' table. 
- [x] Create Model and Resource Controller for the 'session' table. `0-1 Hour`
- [x] Learn the concepts of  Polymorphic relationship. `3-4 Hour`
- [x] Define Polymorphic relation between Session, CodeTrekApplicant and pivot table 'model_has_session'. `1-2 Hour`
- [x] Create view file to implement presentation details page. `0-1 Hour`
- [x] Implement add new presentation - by creating modal and form. `0-1 Hour`
- [x] Implement edit feature - by creating modal and form. `0-1 Hours`
- [x] Implement sorting. `0-1 Hours`
- [x] Implement delete functionality. `1-2 Hours`
- [x] Read about implementing backend validation (Form Requests). `0-1 Hour`
- [x] Create SessionRequest validation.  `0-1 Hour`
- [x] Show appropriate error in frontend incase any validation rules gets failed. `1-2 Hours`
- [x] Learn the concepts of ajax to implement the toast message for creation, updation, deletion and for validation without reloading the page.`4-5 Hours`

## Checklist:

- [x] The PR title follows this syntax: <#IssueID> <PR Title>
- [x] I have linked the issue id in the PR .
- [x] I have performed a self-review of my own code.
- [x]  I have added comments on my code changes where required.

### CODE REVIEW

File: Modules/CodeTrek/Entities/CodeTrekApplicant.php [hunk 1]
```diff
@@ -44,4 +45,9 @@ public function mentor()
     {
         return $this->belongsTo(User::class, 'mentor_id');
     }
+
+    public function sessions()
```
Comment (ImRishabhRawat) [inline Modules/CodeTrek/Entities/CodeTrekApplicant.php:49]: Defines the many to many (Polymorphic relation) between Session and CodeTrekApplicant Model.

File: Modules/Session/Entities/Session.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+class Session extends Model
+{
+    protected $fillable = ['topic_name', 'date', 'link', 'level', 'summary'];
+
+    use HasFactory;
+
+    public function codetrekApplicants()
+    {
+        return $this->morphedByMany(CodeTrekApplicant::class, 'model', 'model_has_sessions');
```
Comment (ImRishabhRawat) [inline Modules/Session/Entities/Session.php:17]: Defines the many to many (Polymorphic relation) between CodeTrekApplicant Model and pivot table "model_has_sessions".

File: Modules/CodeTrek/Entities/CodeTrekApplicant.php [hunk 3]
```diff
@@ -44,4 +45,9 @@ public function mentor()
...
+    {
+        return $this->morphToMany(Session::class, 'model', 'model_has_sessions');
```
Comment (ImRishabhRawat) [inline Modules/CodeTrek/Entities/CodeTrekApplicant.php:51]: Defines the many to many (Polymorphic relation) between Session model and pivot table 'model_has_sessions'.

File: Modules/Session/Resources/assets/js/app.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			.find(".text-danger")
+			.addClass("d-none")
+			.empty();
+	});
+
+	$("#sessionForm, #editSessionForm").on("submit", function(e) {
+		e.preventDefault(); //This stops the page from reloading when forms are submitted.
+		let modalId = $(this)
+			.closest(".modal") // It finds the closest element with the class "modal" from the form
+			.attr("id"); // retrieves its ID attribute
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:16]: These lines of code prevent the page from reloading when forms are submitted, and finds the modal and its id attribute

File: Modules/Session/Resources/assets/js/app.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+				$("#" + modalId).modal("hide");
+				if (modalId === "createSessionModal") {
+					//Checks if the current modal is create modal
+					Vue.$toast.success("Session Created successfully!"); // Display the Created toast message
+				} else {
+					Vue.$toast.success("Session Updated successfully!"); // Display the Updated toast message
+				}
+				location.reload(true);
+			},
+			error: function(response) {
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:37]: This function will run if there was an error.

File: Modules/Session/Resources/assets/js/app.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+				// This function will run if there was an error
+				$(spinnerId).addClass("d-none"); //Disable the visibility of the spinner
+
+				if (response.responseJSON && response.responseJSON.errors) {
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:41]: If a JSON object with the field "errors" is included in the response data, it indicates that there were validation errors.

File: Modules/Session/Resources/assets/js/app.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+					//If a JSON object with the field "errors" is included in the response data, it indicates that there were validation errors.
+					let errors = response.responseJSON.errors;
+
+					form
+						.find(".text-danger")
+						.addClass("d-none")
+						.empty(); //These line find the div in form which arre used to display the error message
+
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:49]: These line find the div of class = 'text-danger' in form which are used to display the error message

File: Modules/Session/Resources/assets/js/app.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+					} else {
+						$("#" + modalId + " #sessionLevelError").addClass("d-none");
+					}
+
+					if (errors.summary) {
+						let text = errors.summary[0];
+						$("#" + modalId + " #sessionSummaryError")
+							.html(text)
+							.removeClass("d-none");
+						Vue.$toast.error("There is some problem in summary");
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:96]: In above, These if conditions checks the error in  a particular field and display the corresponding errors right below the input field and also display a toast message.

File: Modules/Session/Resources/assets/js/app.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+					} else {
+						$("#" + modalId + " #sessionSummaryError").addClass("d-none");
+					}
+
```
Comment (ImRishabhRawat) [inline Modules/Session/Resources/assets/js/app.js:100]: In above, These if conditions checks the error in a particular field related to validation and display the corresponding errors right below the input field and also display a toast message.

**Comment by GauravGusain98** [comment]
This is obsolete. The requirement needs to be revisited

**Pull request closed without merging**